src/particles.zig

```diff
@@ -0,0 +1,441 @@
+			const intPos: vec.Vec4i = @intFromFloat(@floor(positionf64));
+			const light: [6]u8 = main.renderer.mesh_storage.getLight(intPos[0], intPos[1], intPos[2]) orelse @splat(0);
+			const compressedLight =
+				@as(u32, light[0] >> 3) << 25 |
+				@as(u32, light[1] >> 3) << 20 |
+				@as(u32, light[2] >> 3) << 15 |
+				@as(u32, light[3] >> 3) << 10 |
+				@as(u32, light[4] >> 3) << 5 |
+				@as(u32, light[5] >> 3);
+			particle.light = compressedLight;
+
+			i += 1;
+		}
+		previousPlayerPos = playerPos;
+	}
+
+	fn addParticle(typ: u32, pos: Vec3d, vel: Vec3f, collides: bool) void {
+		const lifeTime = properties.lifeTimeMin + random.nextFloat(&seed)*properties.lifeTimeMax;
+
+		particles[particleCount] = Particle{
+			.posAndRotation = vec.combine(@as(Vec3f, @floatCast(pos - previousPlayerPos)), 0),
+			.typ = typ,
+		};
+		particlesLocal[particleCount] = ParticleLocal{
+			.velAndRotationVel = vec.combine(vel, properties.rotVelMin + random.nextFloatSigned(&seed)*properties.rotVelMax),
+			.lifeVelocity = 1/lifeTime,
+			.collides = collides,
+		};
+		particleCount += 1;
+	}
+
+	pub fn render(projectionMatrix: Mat4f, viewMatrix: Mat4f, ambientLight: Vec3f) void {
+		particlesSSBO.bufferSubData(Particle, &particles, particleCount);
+
+		pipeline.bind(null);
+
+		const projectionAndViewMatrix = Mat4f.mul(projectionMatrix, viewMatrix);
+		c.glUniformMatrix4fv(uniforms.projectionAndViewMatrix, 1, c.GL_TRUE, @ptrCast(&projectionAndViewMatrix));
+		c.glUniform3fv(uniforms.ambientLight, 1, @ptrCast(&ambientLight));
+
+		const billboardMatrix = Mat4f.rotationZ(-game.camera.rotation[2] + std.math.pi*0.5)
+			.mul(Mat4f.rotationY(game.camera.rotation[0] - std.math.pi*0.5));
+		c.glUniformMatrix4fv(uniforms.billboardMatrix, 1, c.GL_TRUE, @ptrCast(&billboardMatrix));
+
+		c.glActiveTexture(c.GL_TEXTURE0);
+		ParticleManager.textureArray.bind();
+		c.glActiveTexture(c.GL_TEXTURE1);
+		ParticleManager.emissionTextureArray.bind();
+
+		c.glBindVertexArray(chunk_meshing.vao);
+
+		for(0..std.math.divCeil(u32, particleCount, chunk_meshing.maxQuadsInIndexBuffer) catch unreachable) |_| {
+			c.glDrawElements(c.GL_TRIANGLES, @intCast(particleCount*6), c.GL_UNSIGNED_INT, null);
+		}
+	}
+
+	pub fn getParticleCount() u32 {
+		return particleCount;
+	}
+};
+
+pub const EmitterProperties = struct {
+	gravity: Vec3f = @splat(0),
+	drag: f32 = 0,
+	velMin: f32 = 0,
+	velMax: f32 = 0,
+	rotVelMin: f32 = 0,
+	rotVelMax: f32 = 0,
+	lifeTimeMin: f32 = 0,
+	lifeTimeMax: f32 = 0,
+};
+
+pub const EmitterShapeEnum = enum(u8) {
+	point,
+	sphere,
+	cube,
+};
+pub const DirectionModeEnum = enum(u8) {
+	spread,
+	scatter,
+	direction,
+};
+
+pub const Emitter = struct {
+	shapeType: EmitterShapeEnum,
+	directionMode: DirectionModeEnum = .spread,
+	count: u32,
+	size: f32 = 0,
+	dir: Vec3f = @splat(0),
+	id: []const u8,
+	collides: bool,
+
+	pub fn spawn(self: Emitter, pos: Vec3d) void {
```

Comment:
I would prefer a more precise name 
```suggestion
	pub fn spawnParticles(self: Emitter, pos: Vec3d) void {
```